webrtc/RTCPeerConnection-onnegotiationneeded.html

```diff
@@ -0,0 +1,287 @@
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.createDataChannel('test');
+    pc.close();
+    return awaitNegotiation(pc);
+
+  }, 'negotiationneeded event should not fire if connection is closed');
+
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.createDataChannel('foo');
+
+    return awaitNegotiation(pc)
+      .then(({nextPromise}) => {
+      pc.createDataChannel('bar');
+      return nextPromise;
+    });
+  }, 'calling createDataChannel twice should fire negotiationneeded event once');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To check if negotiation is needed
+      3.  For each transceiver t in connection's set of transceivers, perform
+          the following checks:
+          1.  If t isn't stopped and isn't yet associated with an m= section
+              according to [JSEP] (section 3.4.1.), return "true".
+
+    5.1.  RTCPeerConnection Interface Extensions
+      addTransceiver
+        9.  Update the negotiation-needed flag for connection.
+   */
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+    pc.addTransceiver('audio');
+    return awaitNegotiation(pc);
+  }, 'addTransceiver() should fire negotiationneeded event');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To update the negotiation-needed flag
+      4.  If connection's [[needNegotiation]] slot is already true, abort these steps.
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.addTransceiver('audio');
+    return awaitNegotiation(pc)
+    .then(({nextPromise}) => {
+      pc.addTransceiver('video');
+      return nextPromise;
+    });
+  }, 'Calling addTransceiver() twice should fire negotiationneeded event once');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To update the negotiation-needed flag
+      4.  If connection's [[needNegotiation]] slot is already true, abort these steps.
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    pc.createDataChannel('test');
+    return awaitNegotiation(pc)
+    .then(({nextPromise}) => {
+      pc.addTransceiver('video');
+      return nextPromise;
+    });
+  }, 'Calling both addTransceiver() and createDataChannel() should fire negotiationneeded event once');
+
+  /*
+    4.7.3.  Updating the Negotiation-Needed flag
+      To update the negotiation-needed flag
+      2.  If connection's signaling state is not "stable", abort these steps.
+   */
+  test_never_resolve(t => {
+    const pc = new RTCPeerConnection();
+    const promise = awaitNegotiation(pc);
+
+    return pc.createOffer()
+    .then(offer => pc.setLocalDescription(offer))
+    .then(() => {
+      assert_equals(pc.signalingState, 'have-local-offer');
+      pc.createDataChannel('test');
+      return promise;
+    });
+  }, 'negotiationneeded event should not fire if signaling state is not stable');
```

Comment:
> This is failing for the wrong reason in Firefox Nightly, but I don't see a way to create a track to call pc.addTrack() with, without going through `navigator.mediaDevices.getUserMedia({audio: true, video: true})`

One way to suppress that error in Firefox is by adding audio/video option in createOffer, e.g. `createOffer({ offerToReceiveAudio })`. But I don't think the tests should all use awkward parameters just to work around current browser bugs. Other than the methods being tested, the other methods should be assumed to compliant with the spec, otherwise they should fail the tests without unnecessary assertions.